centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala

```diff
@@ -65,30 +63,130 @@ case class WorkflowFlatMetadata(value: Map[String, JsValue]) extends AnyVal {
       case o: JsString if stripQuotes(cacheSubstitutions).startsWith("~~") =>
         val stripped = stripQuotes(cacheSubstitutions).stripPrefix("~~")
         (!stripQuotes(o.toString).contains(stripped)).option(s"Actual value ${o.toString()} does not contain $stripped")
-      case o: JsString => (cacheSubstitutions != o.toString).option(s"expected: $cacheSubstitutions but got: $actual")
+      case o: JsString =>
+        (cacheSubstitutions != o.toString).option(s"expected: $cacheSubstitutions but got: $actual")
       case o: JsNumber =>
-        (expected != JsString(o.value.toString)).option(s"expected: $cacheSubstitutions but got: $actual")
+        expected match {
+          case JsNumber(value) =>
+            (value.compare(o.value) != 0).option(s"expected: $cacheSubstitutions but got: $actual")
+          case _ => Option(s"metadata $actual is a number, but expected to be the same type as $expected")
+        }
       case o: JsBoolean =>
-        (expected != JsString(o.value.toString)).option(s"expected: $cacheSubstitutions but got: $actual")
+        expected match {
+          case JsBoolean(value) => (value != o.value).option(s"expected: $cacheSubstitutions but got: $actual")
+          case _ => Option(s"metadata $actual is a boolean, but expected to be the same type as $expected")
+        }
       case o: JsArray if stripQuotes(cacheSubstitutions).startsWith("~>") =>
         val stripped = stripQuotes(cacheSubstitutions).stripPrefix("~>")
         val replaced = stripped.replaceAll("\\\\\"", "\"")
-        (replaced != o.toString).option(s"expected: $cacheSubstitutions but got: $actual")
-      case o: JsArray => (expected != JsString(o.toString)).option(s"expected: $cacheSubstitutions but got: $actual")
+        val expectedArray = stringToArray(replaced)
+        expectedArray match {
+          case Success(array) =>
+            (array.elements != o.elements).option(s"expected: $array but got: $actual")
+          case Failure(_) => Option(s"metadata $actual is an array, but expected to be the same type as $replaced")
+        }
+      case o: JsArray =>
+        expected match {
+          case JsArray(elements) =>
+            (elements != o.elements).option(s"expected: $cacheSubstitutions but got: $actual")
+          case _ => Option(s"metadata $actual is an array, but expected to be the same type as $expected")
+        }
       case JsNull => (expected != JsNull).option(s"expected: $cacheSubstitutions but got: $actual")
       case _ => Option(s"expected: $cacheSubstitutions but got: $actual")
     }
 
     matchError.map(s"Metadata mismatch for $key - " + _)
   }
+
```

Comment:
You've done an admirable job creating your own string-to-array parser here, but I wonder if there's anything available in spray that can do this for us. Nice to outsource JSON serialization and deserialization wherever we can.

Reply:
Yup, spray has a `parseJson` method which was able to do this for us!